### PR TITLE
[Core] Refine deprecation procedure and deprecate TStringLong

### PR DESCRIPTION
**Pull request by dpiparo**

### CODE REVIEW

File: core/base/inc/LinkDef2.h [hunk 1]
```diff
@@ -118,7 +118,6 @@
 #pragma link C++ class TString-!;
 //#pragma link C++ class TString::Rep_t-!;
 #pragma link off class TString::Rep_t;
-#pragma link C++ class TStringLong-;
```
Comment (Axel-Naumann) [inline core/base/inc/LinkDef2.h]: Why can we remove this?

File: core/base/inc/TStringLong.h [hunk 2]
```diff
@@ -50,6 +52,8 @@ class TStringLong : public TString {
    virtual Int_t    Sizeof() const;
 
    ClassDef(TStringLong,1)  //Long string class (more than 255 chars)
-};
+} R__SUGGEST_ALTERNATIVE("TString");
```
Comment (Axel-Naumann) [inline core/base/inc/TStringLong.h]: Shouldn't this be `R__ALWAYS_SUGGEST_ALTERNATIVE` and we get rid of the `_LOCAL` part?

File: README/ReleaseNotes/v618/index.md [hunk 3]
```diff
@@ -40,6 +40,9 @@ The following people have contributed to this new version:
 
 ## Deprecation and Removal
 
+## Preprocessor deprecation macros
+  * `R__SUGGEST_ALTERNATIVE("Suggestion text")` macro allows to suggest alternatives to classes. It must be used after the class definition and before the final semicolon. It is activated by the preprocessor defines `R__SUGGEST_NEW_INTERFACE` and `R__SUGGEST_NEW_INTERFACE_LOCAL`. The former variable is useful when deprecation needs to be activated at global level, for example for an entire project, while the latter is useful when defined and undefined immediately before and after an individual class.
+
 ### I/O Libraries
```
Comment (hageboeck) [inline README/ReleaseNotes/v618/index.md]: Should we include a small code example?

```
#ifndef INCLUDE_GUARD
#define R__SUGGEST_NEW_INTERFACE_LOCAL

class DoNotUse {
} R__SUGGEST_ALTERNATIVE("Use ... instead.");
#undef R__SUGGEST_NEW_INTERFACE_LOCAL

#endif INCLUDE_GUARD
```

Comment (dpiparo) [reply README/ReleaseNotes/v618/index.md]: can u add it via github?

File: core/base/inc/LinkDef2.h (hunk 1 above)
Comment (dpiparo) [reply core/base/inc/LinkDef2.h]: We cannot, you are right!

File: core/base/inc/TStringLong.h (hunk 2 above)
Comment (dpiparo) [reply core/base/inc/TStringLong.h]: but how would you avoid having deprecation warnings for all deprecated classes?

**Comment by hageboeck** [comment]
@dpiparo The downside of entering the changes on github is that they disappear when you force-push.

File: core/base/inc/TStringLong.h (hunk 2 above)
Comment (Axel-Naumann) [reply core/base/inc/TStringLong.h]: I probably misunderstand what this is supposed to do.

Right now, whatever `#include`s `ROOT/RConfig.hxx` first, gets do define what `R__SUGGEST_NEW_INTERFACE_LOCAL` is and which effect it has.

`TStringLong.h` defines it around the class definition, unconditionally - whatever users do.

I guessed that the intent is that `TSringLong` always warns, on any use. In that case we could remove `#define ` / `#undef R__SUGGEST_NEW_INTERFACE_LOCAL`, and simply use a `R__ALWAYS_SUGGEST_ALTERNATIVE` that always complains, no matter what's the value of `R__SUGGEST_NEW_INTERFACE`.

**Comment by dpiparo** [comment]
my punishment will be to add the changes myself...

File: README/ReleaseNotes/v618/index.md (hunk 3 above)
Comment (hageboeck) [reply README/ReleaseNotes/v618/index.md]: Done. Don't force-push from now on.

File: core/base/inc/TStringLong.h (hunk 2 above)
Comment (hageboeck) [reply core/base/inc/TStringLong.h]: That might actually work. You would go into hell, though, when you flag an often-used class with this.

Comment (dpiparo) [reply core/base/inc/TStringLong.h]: @Axel-Naumann now I see, it makes sense. One of the intent of this PR was really to see the deprecation in action from another perspective to see how to improve, and it's working!
Changes coming (not force-pushed ;-) )

**Comment by dpiparo** [comment]
@Axel-Naumann : can this go in?

**Pull request merged**